raw-ipa-lib/src/user.rs

```diff
@@ -68,18 +69,46 @@ impl User {
         Ok(())
     }
 
-    fn point_from_matchkey(mk: &[u8; 32]) -> RistrettoPoint {
-        // Note that ristretto wants 64 bytes of input; also we don't know if the input is uniform.
-        // TODO: Consider salting this input somehow (with the provider, perhaps).
-        //       The caveat being that anything we do needs to be standardized.
-        RistrettoPoint::hash_from_bytes::<Sha512>(&mk[..])
+    /// This part ultimately needs to be standardized.
+    /// This should do for now in terms of getting a good input.
+    fn point_from_matchkey(provider: &str, mk: &[u8]) -> RistrettoPoint {
+        let mut input = Vec::with_capacity(2 + provider.len() + mk.len());
+        input.push(u8::try_from(provider.len()).unwrap());
+        input.extend_from_slice(provider.as_bytes());
+        input.push(u8::try_from(mk.len()).unwrap());
+        input.extend_from_slice(mk);
+        RistrettoPoint::hash_from_bytes::<Sha512>(&input)
     }
 
-    pub fn set_matchkey(&mut self, provider: &str, mk: &[u8; 32]) {
-        let m = Self::point_from_matchkey(mk);
+    pub fn set_matchkey(&mut self, provider: impl AsRef<str>, mk: impl AsRef<str>) {
+        let m = Self::point_from_matchkey(provider.as_ref(), mk.as_ref().as_bytes());
         let emk = self.threshold_key.encrypt(m, &mut thread_rng());
-        self.encrypted_match_keys
-            .insert(String::from(provider), emk);
+        trace!("set matchkey for '{}' to '{:?}'", provider.as_ref(), emk);
+        if let Some(old) = self
+            .encrypted_match_keys
+            .insert(String::from(provider.as_ref()), emk)
+        {
+            trace!("old matchkey for '{}' was '{:?}'", provider.as_ref(), old);
+        }
+    }
+
+    /// Deterministically generate a fallback matchkey for a provider that
+    /// hasn't set one.  This uses the fallback secret (`fallback_prk`) generated
+    /// when the user was created.
+    fn fallback_matchkey(&self, provider: &str) -> Ciphertext {
```

Comment:
This function signature uses `&str` intead of `impl AsRef<str>`, why is that?

Reply:
At this point it's not a public API and I don't need to add the generic.  Generics add a small overhead, so I don't want to do that.